PeakNetwork/Core/DecodeInputOperation.swift

```diff
@@ -0,0 +1,242 @@
+            return .failure(SerializationError.noData)
+        }
+        
+        return Result {
+            return (try decoder.decode(D.self, from: data), input.urlResponse)
+        }
+    }
+}
+
+/// Decode a network response into a platform-specific Image type
+open class ImageDecodeOperation: DecodeInputOperation<NetworkResponse, PeakImage> {
+    
+    open override func decode(input: NetworkResponse) -> Result<PeakImage> {
+        if let data = input.data, let image = PeakImage(data: data) {
+            return .success(image)
+        } else {
+            return .failure(ImageResponseOperationError.invalidData)
+        }
+    }
+
+    public enum ImageResponseOperationError: Error {
+        /// Could not initialize the image from the specified data.
+        case invalidData
+    }
+}
+
+
+
+/// Perform two operations, passing the result of the first to the second.
+/// The input to this operation is used as the input to the first operation.
+/// The output of the second operation is used as the output of this operation.
+open class SequenceOperation<First, Second>: ConcurrentOperation, ProducesResult
+    where
+        First: ConcurrentOperation,
+        Second: ConcurrentOperation,
+        First: ProducesResult,
+        First: ConsumesResult,
+        Second: ProducesResult,
+        Second: ConsumesResult,
+        First.Output == Second.Input {
+    
+    public var input: Result<First.Input> = Result { throw ResultError.noResult }
+    public var output: Result<Second.Output> = Result { throw ResultError.noResult }
+
+    internal let queue = OperationQueue()
+    
+    internal let first: First
+    internal let second: Second
+    
+    
+    /// Create a new `SequenceOperation`.
+    ///
+    /// - Parameters:
+    ///   - input: An optional input value.
+    ///   - first: The first operation to perform.
+    ///   - second: The second operation to perform, passed the output of `first`.
+    public init(input: First.Input? = nil, do first: First, passResultTo second: Second) {
+        self.first = first
+        self.second = second
+        if let input = input {
+            first.input = .success(input)
+        }
+    }
+    
+    open override func execute() {
+        guard !isCancelled else { return finish() }
+        second.addResultBlock { [weak self] result in
+            guard let strongSelf = self else { return }
+            if !strongSelf.isCancelled {
+                strongSelf.output = result
+            }
+            strongSelf.finish()
+        }
+        
+        first
+            .passesResult(to: second)
+            .enqueue(on: queue)
+    }
+
+    
+    open override func cancel() {
+        queue.cancelAllOperations()
+        super.cancel()
+    }
+}
+
+/// `DecodableOperation` will attempt to parse the response into a `Decodable` type.
+open class DecodableOperation<D: Decodable>: SequenceOperation<NetworkOperation, JSONDecodeOperation<D>> {
```

Comment:
I understand why these have been left in, but they should probably be deprecated with a view to be removed entirely in a future release of PeakNetwork.

Reply:
Good idea.